src/Simplex/Chat/Library/Commands.hs

```diff
@@ -3536,6 +3517,41 @@ cleanupManager = do
       forM_ contacts $ \ct ->
         withStore (\db -> deleteContactWithoutGroups db user ct)
           `catchChatError` (toView . CRChatError (Just user))
+    cleanupDeletedGroups user@User {userId} = do
+      vr <- chatVersionRange
+      groups <- withStore' $ \db -> getDeletedGroups db vr user
+      forM_ groups $ \g ->
+        cleanupGroup vr g `catchChatError` (toView . CRChatError (Just user))
+      where
+        cleanupGroup :: VersionRangeChat -> Group -> CM ()
+        cleanupGroup vr Group {groupInfo = gInfo, members} = do
+          let contactIds = mapMaybe memberContactId members
+          (errs1, (errs2, deletedCtIds, connIds)) <- lift $ second unzip3 . partitionEithers <$> withStoreBatch (\db -> map (deleteUnusedContact db) contactIds)
+          let errs = errs1 <> mapMaybe (fmap ChatErrorStore) errs2
+          unless (null errs) $ toView $ CRChatErrors (Just user) errs
+          deleteAgentConnectionsAsync user $ concat connIds
+          let deletedCtIds' = catMaybes deletedCtIds
+              members' = map (updateContactId deletedCtIds') members
+          withFastStore' $ \db -> deleteGroupMembers db user gInfo members'
+          withFastStore' $ \db -> deleteGroup db user gInfo
+          where
+            deleteUnusedContact :: DB.Connection -> ContactId -> IO (Either ChatError (Maybe StoreError, Maybe ContactId, [ConnId]))
+            deleteUnusedContact db contactId = runExceptT . withExceptT ChatErrorStore $ do
+              ct <- getContact db vr user contactId
+              ifM
+                ((directOrUsed ct ||) . isJust <$> liftIO (checkContactHasGroups db user ct gInfo))
+                (pure (Nothing, Nothing, []))
+                (deleteCt ct)
```

Comment:
Compared to old code, contacts are deleted before group and members records are removed. So if app is killed mid cleanup, group would still be checked for unused contacts.

Also instead of marking contact for deletion, contact is deleted immediately, without deleting its profile and ldn, as they will be deleted in next step when deleting members.